server/task/task-memory/src/main/scala/org/apache/james/task/eventsourcing/Events.scala

```diff
@@ -20,22 +20,23 @@ package org.apache.james.task.eventsourcing
 
 import org.apache.james.eventsourcing.{Event, EventId}
 import org.apache.james.task.Task.Result
-import org.apache.james.task.{Hostname, Task}
+import org.apache.james.task.TaskExecutionDetails.AdditionalInformation
+import org.apache.james.task.{Hostname, Task, TaskType}
 
 sealed abstract class TaskEvent(aggregateId: TaskAggregateId, val eventId: EventId) extends Event {
   override def getAggregateId: TaskAggregateId = aggregateId
 }
 
-sealed abstract class TerminalTaskEvent(aggregateId: TaskAggregateId, override val eventId: EventId) extends TaskEvent(aggregateId, eventId)
+sealed abstract class TerminalTaskEvent(aggregateId: TaskAggregateId, override val eventId: EventId, taskType: TaskType, additionalInformation: Option[AdditionalInformation]) extends TaskEvent(aggregateId, eventId)
 
 case class Created(aggregateId: TaskAggregateId, override val eventId: EventId, task: Task, hostname: Hostname) extends TaskEvent(aggregateId, eventId)
 
 case class Started(aggregateId: TaskAggregateId, override val eventId: EventId, hostname: Hostname) extends TaskEvent(aggregateId, eventId)
 
 case class CancelRequested(aggregateId: TaskAggregateId, override val eventId: EventId, hostname: Hostname) extends TaskEvent(aggregateId, eventId)
 
-case class Completed(aggregateId: TaskAggregateId, override val eventId: EventId, result: Result) extends TerminalTaskEvent(aggregateId, eventId)
+case class Completed(aggregateId: TaskAggregateId, override val eventId: EventId, result: Result, taskType: TaskType, additionalInformation: Option[AdditionalInformation]) extends TerminalTaskEvent(aggregateId, eventId, taskType, additionalInformation)
```

Comment:
I don't understand this `TaskType` thing: any Task aggregate starts with a `Created` event which contains the task. I guess Task knows its type. So why next events need to repeat this information as the aggregate already knows it?

Reply:
I had a look, i think it's possible but it means to refactor the `DomainObjectConverter` functional interface from : 
```
    public interface DomainObjectConverter<T, U extends DTO> {
        T convert(U dto);
    }
```

to 
```
    public interface DomainObjectConverter<T, U extends DTO, V> {
        T convert(U dto, V context);
    }
```
Those changes are quite pervasive. WDYT ?

Reply:
Can you explain a bit more the rational behind "not having the task type in the Event leads to not being able to deserialize domain objects"?